Properly escape headers in MySQL fast import

clayliddell opened the pull request:
fixes #3606

This PR adds a one-character prefix (`_`) to incoming column names that are not compatible with mysql. These include:

- Numeric column names
- Column names matching any [MySQL reserved words](https://dev.mysql.com/doc/refman/5.7/en/keywords.html).

This is essentially a stopgap, as [Drupal 9 takes care of this for us](https://www.drupal.org/project/drupal/issues/2986452). Once we drop support for Drupal 8 (don't worry no plans for this yet) we can revert this.  

## QA Steps

- [x] Ensure you're on the latest version of `GetDKAN/datastore`.
  - [x] Create a dataset with a resource file which has column names starting with integers.
  - [x] Ensure you're able to import the newly created dataset.
  - [x] Ensure all of the columns in the newly created dataset have a `d_` prefix and have been properly escaped.
- [x] Switch to the `34-allow-column-names-for-table-fields` branch on `GetDKAN/datastore`.
  - [x] Create a dataset with a resource file which has column names starting with integers.
  - [x] Ensure you're able to import the newly created dataset.
  - [ ] Ensure all of the columns in the newly created dataset have their original table names.

## Known issues

Test coverage for the mysql importer is still incomplete. This module will remain marked as "experimental".

Commit:
Properly escape headers in MySQL fast import.

grugnog commented:
I tested this on my local and it's working for me!

grugnog reviewed (changes requested):
While this works (and unblocks users for now), we shouldn't be building SQL through concatenation (as in getSqlStatement) since it is vulnerable to SQL injection. I think we should fix that here by switching to $database->query() and [placeholder arrays](https://www.drupal.org/docs/drupal-apis/database-api/static-queries#placeholder-arrays) for the headers, which will ensure they are properly quoted and escaped.

Commit:
Prefix table field and store original column name.

Commit:
Minor bug fix; updating documentation.

clayliddell commented:
This PR and https://github.com/GetDKAN/datastore/pull/35 complement each other. They should be tested both together, and separately.

clayliddell commented:
This is a non-BC-breaking implementation of the "option b" suggested by @grugnog (using a `d_` prefix to allow columns starting with an integer).

The https://github.com/GetDKAN/datastore/pull/35 PR which I suggested be tested with this PR adds support for preview column names, which differ from SQL table column names, and are derived from the header row of the CSV file provided for a dataset.

### Update

I just found a dataset that preview column names are not being generated properly for:
https://people.sc.fsu.edu/~jburkardt/data/csv/hurricanes.csv

Commit:
Ensure numeric table headers are preserved.

Swap out `array_merge()` for `array_replace()` to ensure numeric indexes are not reset to being zero-indexed.

grugnog commented on modules/datastore/modules/datastore_mysql_import/src/Service/MysqlImport.php:
```diff
@@ -61,26 +71,74 @@ private function getDatabaseConnectionCapableOfDataLoad() {
+      // parsing issues.
+      //
+      // @see https://github.com/GetDKAN/dkan/issues/3606
+      $header = 'd_' . $header;
```
Will this prefix show up in the api & front-end? I wonder if we should be trying to remove this in column names for reads/queries, since it's not really relevant to end users? I also wonder if the regular importer should do add this prefix too for consistency?

grugnog replied:
I should have looked at https://github.com/GetDKAN/datastore/pull/35 first! I think that addresses this for the most part, except perhaps that we may still want to update the regular importer with the same heuristic?

clayliddell replied:
Yes, agreed. The regular importer also seems to be afflicted by the same bug as well.

dafeder replied:
Maybe something for another PR, and we can do this another time, but I think there is consensus that we should bring the datastore library into DKAN, as it so far seems to provide little value outside DKAN and the separation makes development harder.

grugnog commented on modules/datastore/modules/datastore_mysql_import/src/Service/MysqlImport.php:
```diff
@@ -61,26 +71,74 @@ private function getDatabaseConnectionCapableOfDataLoad() {
+    // substitute the truncated characters with a unique hash.
+    if (strlen($column) > self::MAX_COLUMN_LENGTH) {
+      $field = substr($column, 0, self::MAX_COLUMN_LENGTH - 5);
+      $hash = substr(md5($column), 0, 4);
```
Similar question to the d_ prefix here - should this be made transparent to users?

dafeder replied:
Not completely up to date on this issue, but in general, users want control over their columns and column names. If anything comes out different than it went in, people are going to be unhappy.

clayliddell replied:
This particular line pre-existed this PR. It's a constraint of MySQL. MySQL column names can be no longer than 64 characters and this appears to solve that constraint by replacing the last 4 characters with a segment from an md5 hash generated using the whole column name.
Regardless, this PR in conjunction with https://github.com/GetDKAN/datastore/pull/35 solves this issue by using the actual CSV column name provided as the preview column name for datasets while only using the truncated `d_` prefixed column name in MySQL.
TL;DR: the user shouldn't encounter a modified column name with this approach.

grugnog reviewed (commented):
I had a couple of questions about the effect of these changes

grugnog committed:
Prefix only fully-numeric column names.

grugnog commented:
After my comment on https://github.com/GetDKAN/datastore/pull/35#issuecomment-895594753 @dafeder had the idea of only prefixing fully-numeric columns, rather than all of them. Since this has a much more minimal user and API impact (especially since fully-numeric columns aren't working previously) it seems like it can probably go ahead even without https://github.com/GetDKAN/datastore/pull/35 (although that does seem to be working also).

dafeder commented:
It looks like we need to do this for column names that use reserved words (like "index") as well

grugnog committed:
Add MySQL/MariaDB reserved words to prefix rule.

grugnog committed:
Code climate fixed.

grugnog commented:
Since this now has more minimal impacts we are merging without GetDKAN/datastore#35 for now. Still experimental status, so no tests - created #3608 as a follow-up for that.
